cql/src/main/java/org/apache/cassandra/stargate/transport/internal/RequestToHeadersMapper.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright The Stargate Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.stargate.transport.internal;
+
+import java.net.InetSocketAddress;
+import java.util.*;
+
+public class RequestToHeadersMapper {
+  public static final String TENANT_ID_HEADER_NAME = "tenant_id";
+
+  @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+  public static Map<String, String> toHeaders(
+      Optional<InetSocketAddress> sourceAddressWithEncodedTenantId) {
+    Map<String, String> tenantIdHeaders = new HashMap<>();
+    sourceAddressWithEncodedTenantId
+        .map(tenantIdAddress -> UUID.nameUUIDFromBytes(tenantIdAddress.getAddress().getAddress()))
+        .ifPresent(
+            tenantIdUUID -> tenantIdHeaders.put(TENANT_ID_HEADER_NAME, tenantIdUUID.toString()));
```

Comment:
I'm not sure this conversion is correct.

The proxy protocol [handler](https://github.com/stargate/stargate/blob/b04b68424634ca50c1f1fcae4b9aa966f6827c7c/cql/src/main/java/org/apache/cassandra/stargate/transport/internal/HAProxyProtocolDetectingDecoder.java#L73) creates the address as `new InetSocketAddress(msg.sourceAddress(), msg.sourcePort())`, so I believe the tenant ID is the host _name_ portion of this address. The binary address may be different depending on how load balancing is set up.

I'd assume we ought to use UUID.fromString(tenantIsAddress.getHostName())`

@tjake WDYT?

Reply:
In this case we do not even need to parse the UUID string, I think.

Reply:
I think that we need, for example, the IPv6 will look like this: `3ffe:0:0:1:200:f8ff:fe75:50df` whereas the tenant_id should be UUID (with `-`)

Reply:
Is the tenant UUID encoded as a IPv6 address here? If so, then using `UUID.nameUUIDFromBytes()` is not correct.

Reply:
We should add a comment that states that the raw bytes of the IPv6 address are the UUID because it's out of the ordinary and will catch unfamiliar developers off guard.

Reply:
@mpenick what do you mean by `UUID.nameUUIDFromBytes` is not correct.
Based on our discussion on slack, the Ipv6 bytes are UUID. To implement that:
1. We need to extract those bytes, using: `tenantIdAddress.getAddress().getAddress()`
2. Transform bytes to UUID, and this method `UUID.nameUUIDFromBytes` seems to be designed for it.
Could you elaborate more?

Reply:
+1 to documenting and I assume we'd use `new UUID(long mostSigBits, long leastSigBits)` in that case?

Reply:
I've implemented it according to the slack discussion, I've also added a comment describing it. Please take a look.

Reply:
This is very implementation specific logic. Can we add the raw header to the map and deal with this conversion farther down where it's actually used?

Reply:
Our API assumes that we need to pass the generic `Map<String, String>`. Because of that reason, I think we need to transform the address to a `String` format to not complicate the API more. Otherwise, we would need to propagate it as a `byte[]` and we don't want to do it.

Reply:
But can we just pass around the host? Should really be working with publicAddresses rather than tenantIDs.